fix(angular): Use Angular compiler to compile @sentry/angular

Lms24 opened the pull request:
## Background

As brought to attention in #3282, our Angular SDK currently causes a compiler error when including `TraceDirective` (or `TraceModule`) in an Angular application. 

## Problem

The root cause for this problem is that the Angular SDK (as most our SDKs) is currently compiled with `tsc`. While this works fine for regular library interaction, it is a problem for Angular specific SDK items, such as `TraceDirective` or `TraceModule`. The reason is that `tsc` treats these classes (and their decorators) like a normal typescript file and compiles them with the experimental decorators to JS. Angular and the Angular Runtime however, need a specific format which `tsc` does not deliver. 

Upon application compilation, the Angular compiler notices that `TraceDirective` is not compiled correctly and thus raises the error compiler mentioned in the issue. 

## Solution

The solution is to switch from `tsc` to the Angular compiler to compile our Angular SDK. This is done via the Angular CLI and `ng-packagr` which build and package our SDK correctly so that the Angular runtime can process all SDK components (including `TraceDirective`) correctly. 

**Note**: This might be a breaking change as it requires SDK users to use Angular >= 10. Currently, our SDK might support earlier Angular versions despite the fact that we only list Angular 10-13 as peer dependencies of the SDK. With this new compiler, applications using earlier versions than Angular 10 will not be able to compile correctly.

Test:

fixes: #3282 
ref: https://getsentry.atlassian.net/browse/WEB-622

Lms24 commented on packages/angular/lib/src/errorhandler.ts:
```diff
@@ -26,7 +26,7 @@ export interface ErrorHandlerOptions {
   protected readonly _options: ErrorHandlerOptions;
 
-  public constructor(options?: ErrorHandlerOptions) {
+  public constructor(@Inject('errorHandlerOptions') options?: ErrorHandlerOptions) {
```
This `@Inject(...)` Decorator is required by the Angular compiler as the parameter has to be passed to the constructor for DI

Note to myself: check if this is actually true w/ Angular 10

AbhiPrasad commented on packages/angular/angular.json:
```diff
@@ -0,0 +1,28 @@
+{
+  "$schema": "../.././node_modules/@angular/cli/lib/config/schema.json",
```
does this have to be a relative reference to node_modules?

Lms24 replied:
Not in the sense that it would let compilation fail but VSCode doesn't find the schema when only specifying `@angular/cli/lib/config/schema.json`. Is there something different I should try?

AbhiPrasad replied:
No just wondering if there would be problems when the user consumes this package - but I guess we don't include `angular.json` in the final npm asset distribution right?

Lms24 replied:
nope, that's only used for configuring the angular project/compiler

AbhiPrasad commented on packages/angular/ng-package.json:
```diff
@@ -0,0 +1,20 @@
+      "@sentry/utils": "utils"
+    }
+  },
+  "whitelistedNonPeerDependencies": [
```
Think this is called `allowedNonPeerDependencies` now? https://github.com/ng-packagr/ng-packagr/blob/master/docs/dependencies.md#allowing-in-the-dependencies-section

Lms24 replied:
I wanted to change this property to `allowedNonPeerDependencies` but it was introduced in a later version of `ng-packagr` than the one we're using. I could go ahead and try to upgrade `ng-packagr` but I'm pretty sure that new versions supporting `allowedNonPeerDependencies` will not build APF v10 but some newer version which probably breaks Angular 10 compatibility.

AbhiPrasad commented on packages/angular/ng-package.json:
```diff
@@ -0,0 +1,20 @@
+    "tslib",
+    "rxjs"
+  ],
+  "assets": ["README.md"],
```
license is also an asset

Lms24 replied:
right, fixing..

AbhiPrasad commented on packages/angular/ng-package.json:
```diff
@@ -0,0 +1,20 @@
+    "entryFile": "src/index.ts",
+    "umdModuleIds": {
+      "@sentry/browser": "Sentry",
+      "@sentry/utils": "utils"
```
utils shouldn't be exposed globally in umd, they are free to generate their own from `@sentry/utils`.

Taking the `utils` namespace might be problematic behaviour

Lms24 replied:
I'm not entirely sure what to put here to be honest. If I delete the `@sentry/utils` entry, I get the following compiler warning: 

```
WARNING: No name was provided for external module '@sentry/utils' in output.globals – guessing 'utils'
`
``

My understanding is that the `utils` namespace would then still be taken by the compiler's guess. Or should I simply give them a different namespace, say `sentryUtils`?

AbhiPrasad replied:
I found this - https://stackoverflow.com/a/61442340

Let's do something like `Sentry.util` then?

AbhiPrasad commented on packages/angular/angular.json:
```diff
@@ -0,0 +1,28 @@
+      "projectType": "library",
+      "root": ".",
+      "sourceRoot": "src",
+      "prefix": "lib",
```
this prefix still apply?

Lms24 replied:
Good catch, `prefix` is an automatically generated property that tells the Angular CLI which prefix to add to the selector of a new component/directive/pipe when creating one via the CLI. Since we're not using prefixes in our directive selector (the selector is just `trace` instead of e.g. `lib-trace`), we can simply remove the property.

AbhiPrasad reviewed (commented):
quick pass through, when this is properly ready to review lmk and I'll dive in!

Lms24 commented:
Added some information and still open question to the PR description

lobsterkatie commented:
> A more detailed explanation on the compiler differences is given in https://github.com/getsentry/sentry-javascript/pull/4641.

Is this meant to point somewhere else, or is this copy pasta from somewhere else pointing to here?

lobsterkatie commented:
4. The Angular compiler prints a warning when including scripts in the package.json that should be published, due to a [vulnerability](https://blog.npmjs.org/post/141702881055/package-install-scripts-vulnerability). For the Angular SDK this is easily fixable with a flag in the ng-package.json config to simply exclude scripts from the generated package.json. Can/should we exclude them?

Yes. No reason why consumers of the SDK need our scripts.

lobsterkatie commented on packages/angular/package.json:
```diff
@@ -25,24 +25,39 @@
-    "@angular/common": "^10.0.3",
-    "@angular/core": "^10.0.3",
-    "@angular/router": "^10.0.3"
+    "@angular/common": "~10.2.5",
+    "@angular/core": "~10.2.5",
+    "@angular/router": "~10.2.5",
+    "@angular/compiler": "~10.2.5",
+    "@angular-devkit/build-angular": "~0.1002.4",
+    "@angular/cli": "~10.2.4",
+    "@angular/compiler-cli": "~10.2.5",
+    "@types/node": "^12.11.1",
+    "codelyzer": "^6.0.0",
+    "ng-packagr": "^10.1.0",
+    "ts-node": "~8.3.0",
+    "tslint": "~6.1.0",
+    "typescript": "~4.0.2",
+    "zone.js": "~0.10.2"
```
Two questions here:

- Is there a reason we need to `~`-pin these rather than `^`-pinning them? (Not a problem if we need to, just curious.)

- Are all of these direct dev dependencies (as opposed to transitive ones)?

Lms24 replied:
1. The `~`-pins were set by the Angular CLI when the Angular project from which I started out was initialized. What I know is that each Angular version works with different ranges of dependency versions. An example would be that Angular 10.2 requires typescript 3.9.x or 4.0.x according to this compatibility [matrix](https://gist.github.com/LayZeeDK/c822cc812f75bb07b7c55d07ba2719b3). I assume hat the Angular CLI `~`-pins versions where only such specific ranges of versions are compatible with the used Angular version.

2. When I generated the Angular project first via the Angular CLI, it first set up a "outer" `package.json` in which these now dev dependencies were declared as dependencies. The second "inner" directory and`package.json` only contained dependencies required by the actual library and it relied on the outer one to provide all necessary dependencies for an Angular "environment" to work properly (such as the CLIs, compiler, `ng-packagr`, etc.). What I then did to flatten the project structure was to merge these `package.json`s into one which resembled our original one and I therefore put the Angular "environment" dependencies into the `devDependencies` section. This is btw similar to how the original Angular SDK's `package.json` registered (dev-)dependencies. Presumably, the listed dependencies are direct dependencies that are needed to compile an Angular project. However, I might be wrong here as my knowledge on direct vs. transitive dependencies is limited tbh.

Looking at this list though, I will remove `codelyzer` and `tslint` as I'm pretty sure we do not need them with our monorepo (since we're using `eslint`...).

lobsterkatie replied:
As for dependencies, the question I was asking was less "are these used?" (though good catch that the linters aren't) and more "which of these are the ones we actually use directly?" I was thinking if there were extras, they might be sub-dependencies (dependencies of our dependencies), but what you're saying now makes it make more sense where they all came from.

Keep in mind that we don't want the dev dependencies for an angular _project_ (which properly belong in our users' `package.json` files), we want the dev dependencies for an angular _library_, i.e., only the things which it takes to do our work as library developers. My advice would be to clear out that section entirely, and then use yarn to add the things we're actually using back in. The only thing I actually see us using anywhere is the `ng` CLI tool, which makes me think that if you cleared that section out and ran `yarn add -D <whatever-npm-package-provides-that-CLI-tool>`, you might have everything you need.

(It also might be worthwhile having a meta convo about how to deal with test projects and SDK changes in general. Happy to do that whenever.)

Lms24 replied:
Thanks for the idea of the bottom up approach for the dev dependencies. I did that and started with adding back `@angular/cli` (which we need for sure, as you correctly pointed out). What did not happen was that it would install other dependencies that it need as well. But by running `yarn build` I could identify easily which other packages I'd need and I continued adding those as well. At the end, it turned out that we needed most of the compilation and packaging related and Angular core/common/router packages (which makes sense, since we're using parts of these libraries in our code). 
However, I could get rid of `ts-node`, `@types/node` and `zone.js`.

lobsterkatie commented on packages/angular/tsconfig.ngc.json:
```diff
@@ -0,0 +1,26 @@
+{
+  "extends": "./tsconfig.json",
+  "compilerOptions": {
+    "target": "es2015",
+    "lib": [
+      "dom",
+      "es2018"
```
I'm curious about the mismatch of versions here. Does the angular compiler provide polyfills for es2016-es2018 stuff?

Lms24 replied:
Also here this is what was originally set by the Angular CLI.
I think what this basically allows us to do is to theoretically use es2018 but `ngc` compiles the sources to es2015. So yes, I guess it provides polyfills for >es2015 stuff.

lobsterkatie commented on packages/angular/tsconfig.ngc.json:
```diff
@@ -0,0 +1,26 @@
+/* To learn more about this file see: https://angular.io/config/tsconfig. */
```
This is helpful. Could you add similar links to the other new json files?

Lms24 replied:
Good itdea! I added it to `angular.json` and it worked fin. Then I also wanted to add it to `ng-package.json` which gave me a build error. Apparently, `ng-packagr` can't handle JSON with comments.

lobsterkatie replied:
Oh, yeah, `package.json` files can't handle comments (and the npm people have said they never will), so I guess it's not super surprising that `ng-package.json` can't, either.

lobsterkatie commented on packages/angular/tsconfig.ngc.json:
```diff
@@ -0,0 +1,26 @@
+    "skipTemplateCodegen": true,
+    "strictMetadataEmit": true,
+    "enableResourceInlining": true,
+    // As per Angular 10, the recommendation from the library creation guide
+    // is to disable compilation for Ivy in production builds to ensure
+    // compatibility with Angular 10.
+    // For Angular 11-13 applications, ngcc and the Angular linker ensure that
+    // the library code is processed to match the specific Angular runtime version.
```
```suggestion
    // As per Angular 10, the recommendation from the library creation guide
    // is to disable compilation for the Ivy rendering engine in production builds
    // to ensure compatibility with Angular 10.
    // For Angular 11-13 applications, ngcc and the Angular linker ensure that
    // the library code is processed to match the specific Angular runtime version.
```

I'm not sure I quite understand this comment (specifically the second half). Does that mean that for 11-13 it will reenable ivy? Or will ignore this setting all together? (I also made a small edit to the first half just to give a non-angular reader a tiny bit of context here.)

(EDIT: Wow, do they make that hard to parse, with red on red. The change is to replace "Ivy" with "the Ivy rendering engine" and then move "to ensure" onto the next line for line wrapping purposes.)

Lms24 replied:
I can see how this comment is confusing. Thanks for the suggestion, I included it and changed the second half of the explanation.

Lms24 replied:
> Is this meant to point somewhere else, or is this copy pasta from somewhere else pointing to here?

Whoops, this should have linked to #4644. Changed it.

> I think this should now be `./build` rather than `./dist`, right?

Yes, fixed.

> Also, the repo uses yarn as our package manager/script runner, so probably more idiomatic to `s/npm run/yarn`.

Right, also fixed.


> I can't speak to the angular-specific bits, but overall this looks pretty reasonable.

Cool, thanks for the review!

lobsterkatie reviewed (commented):
I can't speak to the angular-specific bits, but overall this looks pretty reasonable.

Lms24 commented:
> Is this meant to point somewhere else, or is this copy pasta from somewhere else pointing to here?

Whoops, this should have linked to #4644. Changed it.

> I think this should now be `./build` rather than `./dist`, right?

Yes, fixed.

> Also, the repo uses yarn as our package manager/script runner, so probably more idiomatic to `s/npm run/yarn`.

Right, also fixed.


> I can't speak to the angular-specific bits, but overall this looks pretty reasonable.

Cool, thanks for the review!

Lms24 commented:
Corrected some minor yarn commands in PR description (`yarn pack` --> `yarn run pack`)

Lonli-Lokli commented on packages/angular/package.json:
```diff
@@ -25,24 +25,34 @@
     "@sentry/utils": "6.18.0",
     "rxjs": "^6.6.0",
-    "tslib": "^1.9.3"
+    "tslib": "^2.0.0"
```
Does it make sense to add Rxjs 7 support as it is with Angular itself? Another question, do you still have node 6 as minimum? Angular drop support for it
https://github.com/angular/angular/blob/master/packages/core/package.json

Lms24 replied:
Hey, thanks for writing in! 
I agree with you that if we just had to support the most recent Angular version, we could bump Rxjs and Node but we want to support Angular 10 and upwards which afaik requires us to stick with Rxjs 6. For Node, Angular 10 doesn't limit Node support, at least not as far as I can see from `package.json` here: https://github.com/angular/angular/blob/10.0.x/packages/core/package.json

If you have a suggestion on how we could handle Angular version support better (v10 upwards), I'd be happy to read it though.

Lonli-Lokli replied:
You can look into Angular\Node matrix here (https://www.redcricket.net/en/news-en/compatibility-list-for-angular-angular-cli-and-node-js/) 
Support for Node 10 has beeen dropped last year https://github.com/angular/angular/pull/41544

Re:Rxjs, I just want to see Rxjs7 be allowed, as now Angular 13 allow using for newer version

Lms24 replied:
I looked into this and according to the matrix, Angular 10 supported Node 10. So I guess we could bump node to Node 10 for @sentry/angular but I'll check with my team first.

Regarding Rxjs, I think we can get away with allowing both, 6 and 7 as peer dependencies rather than rxjs 6 as a direct dependency (which is recommended by the Angular library creation guide anyway). Again, I'll check with the team first.

Lonli-Lokli commented on packages/angular/tsconfig.ngc.json:
```diff
@@ -0,0 +1,28 @@
+  "angularCompilerOptions": {
+    "skipTemplateCodegen": true,
+    "strictMetadataEmit": true,
+    "enableResourceInlining": true,
```
Have you considered more strict settings?
https://angular.io/guide/angular-compiler-options

Lms24 replied:
Up to this point, I went with the default parameters included by the Angular CLI when setting up an Angular library via the CLI (with Angular 10). Do you have suggestions what we should add here (keeping in mind Angular 10-13 support)?

Lonli-Lokli replied:
Have you run this setup with `ng new` or `ng new --strict`?

Specifically, the strict flag does the following:

- Enables strict mode in TypeScript
- Turns template type checking to Strict
- Default bundle budgets have been reduced by ~75%
- Configures linting rules to [prevent declarations of type ](https://palantir.github.io/tslint/rules/no-any/)[any](https://palantir.github.io/tslint/rules/no-any/)
- Configures your app as side-effect free to enable more advanced tree-shaking

Lms24 replied:
I see, indeed I set it up just with `ng new`. I'll look into this and I'll see what we can do on this end. 

Just to be sure though, using the `--strict` flag would just add additional parameters to the tsconfig? And since Angular seems to make a few differences in the tsconfig when setting up libraries instead of apps, are the stricter settings also recommended for library creation? Asking because I find it odd that the Angular library creation guide does not mention them.

Lms24 replied:
Thanks for the links and for the expanations. Some of the TS specific strict settings I see in the linked diff `tsconfig.json` are already part of our general `tsconfig` which also applies to our Angular SDK (see https://github.com/getsentry/sentry-javascript/blob/master/packages/typescript/tsconfig.json). So that's at least something. 

For the remaining ones and Angular specific ones, I'll try to add them and check if they break something on our end. Can't tell you yet if it will happen in this PR or a subsequent one. If you see something missing once we release the new Angular SDK with v7, please feel free to open up an issue. Of course, a PR would be welcome as well (always and not only limited to this aspect of our SDKs) ;)

Lms24 replied:
While checking out the `--strict` flag in `ng-new`, I found that certain options are only applied to angular applications (`ng g application`) and not to libraries (`ng g library`). E.g. budgets.

Some others, such as `strictTemplates: true` apply to both. Gonna include them.

Lms24 commented:
> Have you found `yarn pack` not to work? Generally the `run` is optional.

I thought so, too but `yarn pack` will pack the tarball from the directory it's called whereas calling `yarn run pack` will execute the `pack` command specified in `package.json`. This command is not equal to `yarn pack` anymore as it points to the `package.json` in `build`

lobsterkatie commented:
> > Have you found `yarn pack` not to work? Generally the `run` is optional.
> 
> I thought so, too but `yarn pack` will pack the tarball from the directory it's called whereas calling `yarn run pack` will execute the `pack` command specified in `package.json`. This command is not equal to `yarn pack` anymore as it points to the `package.json` in `build`

Ohhhhh. Because `yarn pack` is an actual thing. I think we should rename the script in that case, to avoid confusion (and to make `yarn newScriptName` work).

Lms24 commented:
I agree, another name for the script would be great (thinking about `yarn package, `yarn tarball` or something similar...) to clear up the confusion. I think this should go into its own issue/task though as it requires changes in many (all?) of our JS SDKs and probably some adjustments in our CI/craft config.

Lms24 commented:
Revisited and updated this PR so that it can be merged into `7.x`. Additionally applied some cleanup and removed the legacy options as we decided to not release a legacy package for the time being.
